src/librbd/crypto/CryptoObjectDispatch.cc

```diff
@@ -79,13 +92,246 @@ struct C_EncryptedObjectReadRequest : public Context {
+                    restart = true;
+                  } else if (r == -EEXIST && !exclusive) {
+                    restart = true;
+                  }
+                  if (restart) {
+                    // remove alignment data
+                    data->substr_of(
+                            data_ref, object_off - aligned_object_off, data_length);
+                    auto req = new C_UnalignedObjectWriteRequest<I>(
+                            image_ctx, crypto, object_no, object_off, data,
+                            cmp_data, mismatch_offset, io_context, op_flags,
+                            write_flags, assert_version, parent_trace,
+                            object_dispatch_flags, journal_tid, on_dispatched);
+                    req->send();
+                  } else {
+                    on_dispatched->complete(r);
+                  }
+                });
+
+        // send back aligned write back to get encrypted and committed
+        auto write_req = io::ObjectDispatchSpec::create_write(
+                image_ctx, io::OBJECT_DISPATCH_LAYER_NONE, object_no,
+                aligned_object_off, std::move(*aligned_data), io_context,
+                op_flags, new_write_flags, version,
+                journal_tid == nullptr ? 0 : *journal_tid, parent_trace, ctx);
+        write_req->send();
+      };
+    }
+
+    void send() {
+      read_req->send();
+    }
+
+    void finish(int r) override {
+      ldout(cct, 20) << "r=" << r << dendl;
+
+      bool no_obj = false;
+      if (r == -ENOENT) {
+        no_obj = true;
+      } else if (r < 0) {
+        onfinish->complete(r);
+        return;
+      }
+
+      if (cmp_data != nullptr) {
+        auto& read_bl = extents.back().bl;
+        std::optional<uint64_t> found_mismatch = std::nullopt;
+        if (no_obj) {
+          read_bl.append_zero(cmp_data->length());
+        } else if (read_bl.length() < cmp_data->length()) {
+          read_bl.append_zero(cmp_data->length() - read_bl.length());
+        } else {
+          ceph_assert(read_bl.length() == cmp_data->length());
+        }
+
+        auto it1 = cmp_data->cbegin();
+        auto it2 = read_bl.cbegin();
+        for (uint64_t idx = 0; idx < cmp_data->length(); ++idx) {
+          if (*it1 != *it2) {
+            found_mismatch = std::make_optional(idx);
+            break;
+          }
+          ++it1;
+          ++it2;
+        }
+
+        if (found_mismatch.has_value()) {
+          if (mismatch_offset != nullptr) {
+            *mismatch_offset = found_mismatch.value();
+          }
+          onfinish->complete(-EILSEQ);
+          return;
+        }
+
+        extents.pop_back();
+      }
+
+      for (auto& extent : extents) {
+        if (extent.offset < object_off) {
+          if (no_obj) {
+            aligned_data->prepend_zero(extent.length);
+          } else {
+            extent.bl.append(*aligned_data);
+            *aligned_data = extent.bl;
```

Comment:
What's the intention here? It's a little confusing at first glance?

Reply:
Basically this prepends the write data (for aligning the write).
bufferlist does not support prepend, so had to split it to append+move.
makes sense? should I add a comment?

Reply:
actually maybe better to use splice? as it will move data instead of copy?

Reply:
`splice` is probably simpler to understand and you can re-use it for the zero and non-zero cases by just running `extent.bl.append_zero(extent.length - extent.bl.lenght())` before the splice.